src/docs/arch-eviction.dox

```diff
@@ -1,15 +1,108 @@
 /*! @arch_page arch-eviction Eviction
 
-Eviction represents the process or removing old data from the cache,
-writing it to disk if it is dirty. The general strategy uses a dedicated
-set of eviction threads that are tasked with identifying candidate data.
-If the data needs to be written, it is reconciled (converting the
-in-memory format to on-disk format), and then written. Clean memory
-can be freed if needed.
-
-Eviction cannot be triggered directly by API calls, but happens as a result
-of enough data being dirtied.
+@section eviction_overall Eviction
+Eviction in WiredTiger is a process to remove data that is currently not being accessed out of the
+memory to free up enough space to read data back into memory that is requested by the user but currently
+reside on the disk. If the data needs to be written, it is reconciled (converting the in-memory format
+to on-disk format), and then written.
 
 Eviction is managed using \c WT_EVICT_QUEUE structures,
 each of which contains a list of \c WT_EVICT_ENTRY structures.
+
+Eviction is composed of one eviction server, several eviction worker threads and three shared
+eviction_queues (two ordinary queues and one urgent queue).  Pages which are being marked to
+evict forcibly due to cache pressure goes on a urgent queue and the pages in the urgent queue
+take precedence over two ordinary queues.
+
+A user can pass the minimum and maximum number of eviction worker threads to be created in the
+configuration string, but however there is a mechanism to dynamically adjust right number of
+eviction worker that gave the best throughput depending on the current eviction load.
+
+- Eviction server walks the pages in the btree to find the candidates to be evicted ,sorts the
+  list of candidates and throws the 1/3 of the candidates to do an approximate LRU and pushes the
+  remaining pages into the eviction_queues.
+- Eviction worker threads goes over the eviction_queues to pop the pages and tries to evict.
```

Comment:
As mentioned, we need more details on how eviction server and candidates interact.

We can talk about how eviction server moves from one file to the other walking the pages trying to find evict-able candidates. Server populates queues while doing so, periodically stopping when the queues are full. Server also tries t be fair in walking - remembering where it was in a tree, and proceeding from there. Also dividing the work of finding candidates between trees, proportional to the size of the trees in memory.

Candidates wait for queues to be populated, working with the server through a CV signalling mechanism. They then get to work independent to the server, till the queues contain enough work for the candidates.

Then we can talk about candidates popping pages off the queues, proceeding with checking if they are still evictable and locking the page if they decide to proceed with eviction.

I am not sure if we decided to write about reconciliation as part of this guide or separate to it. But reconciliation of core to dirty eviction. So we should talk at-least some amount about it here.

Reply:
Explained about the eviction server and worker threads in detail.